sparse/_coo/common.py

```diff
@@ -1152,11 +1253,52 @@ def unique_values(x, /):
     elif not isinstance(x, COO):
         x = x.asformat(COO)
 
-    x = x.flatten()
-    values = np.unique(x.data)
-    if x.nnz < x.size:
-        values = np.sort(np.concatenate([[x.fill_value], values]))
-    return values
+    return x
+
+
+@numba.jit(nopython=True, nogil=True)
+def _sort_coo(
+    coords: np.ndarray,
+    data: np.ndarray,
+    fill_value: float,
+    sort_axis_len: int,
+    descending: bool,
+) -> None:
+    assert coords.shape[0] == 2
+    group_coords = coords[0, :]
+    sort_coords = coords[1, :]
+
+    result_indices = np.empty_like(sort_coords)
+    offset = 0  # tracks where the current group starts
+
+    # iterate through all groups and sort each one of them
+    for unique_val in np.unique(group_coords):
+        # .copy() required by numba, as `reshape` expects a continous array
+        group = np.argwhere(group_coords == unique_val).copy()
+        group = np.reshape(group, -1)
+        group = np.atleast_1d(group)
+
+        # SORT VALUES
+        if group.size > 1:
+            # np.sort in numba doesn't support `np.sort`'s arguments so `stable`
+            # keyword can't be supported.
+            # https://numba.pydata.org/numba-doc/latest/reference/numpysupported.html#other-methods
+            data[group] = np.sort(data[group])
+            if descending:
+                data[group] = data[group][::-1]
+
+        # SORT INDICES
+        fill_value_count = sort_axis_len - group.size
+        indices = np.arange(group.size)
+        # find a place where fill_value would be
+        for pos in range(group.size):
+            if (not descending and fill_value < data[group][pos]) or (descending and fill_value > data[group][pos]):
+                indices[pos:] += fill_value_count
+                break
```

Comment:
Just `argsort` the `values` and add the `first_idx` on top if needed.

Reply:
I think that `argsort` isn't sufficient. In COO format `coords` **must aways be sorted**.
So values must be actually sorted to get indices in the ascending order. If (generally) `coords` aren't sorted it results in an undefined behavior: 
https://github.com/pydata/sparse/blob/82fb0d5d538e5071a017b5ecc0904914b2cedf72/sparse/_coo/core.py#L47-L50
WDYT?

Reply:
`sorted=True` is an optimization to be applied when passing in sorted coords; it will internally sort them anyway if not.

Reply:
Ah, then I need to call `COO` constructor here instead of sorting `coords` inplace (which is not even correct here as I think about it right now, as it modifies passed object instead of operating on a copy)

Reply:
I have one more question:
I don't think that adding the `first_idx` on top if needed is sufficient.

Given `np.argsort` results in `[0,1,2,3]` indices where data is `[0,1,3,4]` it might be possible that we have `fill_value=2` and `first_idx` needs to be added to only specific subset of indices.

In the current implementation it's done simply by finding a specific place where fill values go and using increasing the rest by `indices[pos:] += fill_value_count`.

With `np.argsort` output that is not sorted I need to iterate through all values to check if it's before or after fill values and increase only then.

In the current scenario:
1. We sort values
2. We define indices just as `range(x)`
3. We find a place where `fill_values` go and increase indices by this offset
4. The resulting coords is ready for COO without sorting

In the proposed scenario:
1. We argsort values (obtaining indices)
2. We need to iterate through all values to find those indices that go past `fill_values` and increase by offset
3. The resulting coords still need to be sorted when constructing COO.

I would prefer to stick to the first option, WDYT?

Reply:
That sounds perfectly reasonable to me.